rust/src/producer.rs

```diff
@@ -0,0 +1,182 @@
+use rdkafka::producer::{BaseRecord, ThreadedProducer};
+use rdkafka::producer::{DeliveryResult, ProducerContext};
+use rdkafka::ClientContext;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use thiserror::Error;
+
+/// This structure wraps the parameters to initialize a producer.
+/// This struct is there in order not to expose the rdkafka
+/// details outside.
+#[derive(Debug, Clone)]
+pub struct KafkaConfig {
+    config_map: HashMap<String, String>,
+}
+
+impl KafkaConfig {
+    pub fn new_config<V>(bootstrap_servers: V, override_params: Option<HashMap<String, V>>) -> Self
+    where
+        V: Into<String>,
+    {
+        let mut config_map: HashMap<String, String> = HashMap::new();
+        config_map.insert("bootstrap.servers".to_string(), bootstrap_servers.into());
+
+        let config = Self { config_map };
+
+        apply_override_params(config, override_params)
+    }
+
+    pub fn new_producer_config<V>(
+        bootstrap_servers: V,
+        override_params: Option<HashMap<String, String>>,
+    ) -> Self
+    where
+        V: Into<String>,
+    {
+        let config = KafkaConfig::new_config(bootstrap_servers, None);
+
+        apply_override_params(config, override_params)
+    }
+}
+
+impl From<KafkaConfig> for RdKafkaConfig {
+    fn from(item: KafkaConfig) -> Self {
+        let mut config_obj = RdKafkaConfig::new();
+        for (key, val) in item.config_map.iter() {
+            config_obj.set(key, val);
+        }
+        config_obj
+    }
+}
+
+fn apply_override_params<V>(
+    mut config: KafkaConfig,
+    override_params: Option<HashMap<String, V>>,
+) -> KafkaConfig
+where
+    V: Into<String>,
+{
+    if let Some(params) = override_params {
+        for (param, value) in params {
+            config.config_map.insert(param, value.into());
+        }
+    }
+    config
+}
+
+pub struct CaptureErrorContext;
+
+impl ClientContext for CaptureErrorContext {}
+
+impl ProducerContext for CaptureErrorContext {
+    type DeliveryOpaque = ();
+
+    fn delivery(&self, _: &DeliveryResult, _delivery_opaque: Self::DeliveryOpaque) {
+        // TODO: Do something useful. AS of now guarantees of delivery
+        // in this producer are not particularly critical.
+    }
+}
+
+/// Kafka producer errors.
+#[derive(Error, Debug)]
+pub enum ClientError {
+    /// Failed to send a kafka message.
+    #[error("failed to send kafka message")]
+    SendFailed(#[source] rdkafka::error::KafkaError),
+
+    /// Failed to create a kafka producer because of the invalid configuration.
+    #[error("failed to create kafka producer: invalid kafka config")]
+    InvalidConfig(#[source] rdkafka::error::KafkaError),
+}
+
+/// A basic Kafka Producer trait.
+///
+/// We do not neet to set headers or key for this data.
+pub trait Producer {
+    fn send(&mut self, topic_name: String, payload: &[u8]) -> Result<(), ClientError>;
+}
+
+pub struct DummyProducer<'a> {
+    pub messages: &'a RefCell<Vec<(String, Vec<u8>)>>,
+}
```

Comment:
> Or alternatively Arc<Mutex<…>> if you want it to be threadsafe.

This is indeed only for test, so I don't care about thread safety. Thanks for the `#[cfg(test)]`

Reply:
> In order to avoid the 'a lifetime, you can wrap that up in a Rc.
Or alternatively Arc<Mutex<…>> if you want it to be threadsafe.

I have a question about this, as I am not sure how I would achieve what I do in the accountant test if I used an `Rc`: https://github.com/getsentry/usage-accountant/pull/22/files#diff-b5869fbf23fdb0be35b0bdcc38b0985c55c8c83547ddcdae89c92cd147c9ca06R127.

`DummyProducer` is a mock object to stand in for a real producer. Its goal is to be able to assert what was produced otherwise I would have to run a kafka broker for tests. 
I instantiate `DummyProducer` with a `messages` vector. When the test is done I checl what is in the vector.
Accountant owns the Producer (mutable) and the Producer owns the vector (mutable).

I used a reference to a `RefCell` to be able to access the `messages` field from the test code after the test is done. I don't think I can borrow `messages` after the Accountant is built otherwise as it would be owned by the Producer which is owned by the Accountant.

If I remove the lifetime, I cannot have the structure owning a reference anymore (the reference to the RefCell), that reference is what allowed me to borrow messages at the end of the test even if it is owned (mutable) by the producer.

So it seems to me the lifetime is needed in `DummyProducer` at least. 
Am I missing something or is there a better way to achieve this ?

Reply:
As you correctly said, you have a `messages` "owned" by the current stack frame, and pass a reference of it into the `DummyProducer`, thats why it has a lifetime `'a` tied to the lifetime of the stack frame.

If you replace that with a `Rc<RefCell<…>>` (or `Arc<Mutex<…>>`), you have shared ownership of your `messages` between the current stack frame, and the one you put into the `DummyProducer`.
Thus, the `DummyProducer` has its own "owned" (reference counted) reference of `messages` which is `'static`.

The logic is completely the same, but instead of using a reference to something on the stack, you would have shared ownership of something on the heap, not tied to any lifetimes.

You can have `let messages = Default::default();` (compiler should be able to infer the type), and then `DummyProducer { messages: Rc::clone(&messages) }` to create shared ownership.
The `Rc` will automatically deref to `&RefCell`, so you wouldn’t have to change any of the usage sites.

Reply:
> DummyProducer { messages: Rc::clone(&messages) }

I think this is what I was getting wrong. Thanks. 
Let me try.

Reply:
indeed it worked.